### ISSUE TITLE
CommonJS being compiled with "void 0" and futility of supporting pre-ES5 in general

### ISSUE THREAD

**Issue by revelt** [rollup/rollup-plugin-babel#195]
Hi all,

I'm using a generic setup (`env` preset in `.babelrc`) and generating three builds, ES modules, CommonJS and UMD. Standard stuff.

Looking at my CommonJS build, I noticed that the code is output with `void 0` instead of `undefined`. It looks like the `void 0` was necessary pre-ES5 and now it's [not a current best practice](https://eslint.org/docs/rules/no-void). According to Wikipedia, ES5, the ECMA-262 5th edition was published on December 3, 2009.

I'm wondering, is it still relevant and maybe there are other unneeded "fixes" like this in our CommonJS builds.

In theory, this can be solved customising `.babelrc`, but we should be all using a generic, reasonable settings should come out of the box (and I believe `void 0` is not reasonable). I expected `env` to be that generic config but it's not.

**Comment by xtuc** [comment]
`void 0` is better for generated code because you can't override it and it's smaller than `undefined`. It doesn't change the behavior of the code, I'm not sure to understand what's your issue with it?

**Comment by Andarist** [comment]
You should file issue about this in babel's repo, this is just a plugin for rollup - not responsible for what gets transpiled.

But as `void 0` is more safe and there is no real difference between those 2, I'm not sure what problem would you like to solve here.

Closing this issue as it's filed against wrong repository. Feel free to comment further though.

**Comment by revelt** [comment]
hi guys, thanks for quick response.

> void 0 is better for generated code because you can't override it
@xtuc we _could_ override it pre-ES5, basically 10 years ago. Not any more. `undefined` is safe and perfectly fine as it is.

> I'm not sure what problem would you like to solve here.
@Andarist well, at minimum, redundant actions (like `void 0` thing) waste our precious build resources because original code could have been left untouched as it is, with `undefined`. Speaking about the production code, `void` is an operator which we call on a primitive value, not a primitive value. To me it looks like it's a perf penalty, at least theoretical.

> You should file issue about this in babel's repo
Babel as a tool, will deliver what we ask. If `env` is meant to compile to support pre-ES5, that what it's meant to be doing. Babel as a tool is working correctly. We, on other side, Rollup users, are currently advising `env` in documentation as I'm sure 99% of Rollup users will use it. However,  believe Rollup documentation is wrong advising `env`.

Basically, we should review what gets compiled on our CommonJS builds, find out the official, default Babel config which aims at the new, reasonably lowest bar (not pre-ES5), pack it up as a "recommended" and publish if it's too unwieldy in loose bits, then update all Rollup documentation to use it instead of `env`. This config might be not the best for other users and that's fine. Maybe some people need to compile to pre-ES5, they could continue to use `env`.

Do you see my point?

**Comment by xtuc** [comment]
> original code could have been left untouched as it is

A compiler (Babel) doesn't work like that, we are (re-)generating the entire code anyway, so having `void 0` or `undefined` is the same thing, apart that `void 0` is actually smaller.

I see your point and it's worth opening an issue on the babel-preset-env issue tracker, which is on https://github.com/babel/babel. Thanks.

**Comment by xtuc** [comment]
> To me it looks like it's a perf penalty, at least theoretical.

I don't think it actually hurts perf, `void 0` is two operation but it's guarantee to always have the same result unlike `undefined` (not pre ES5 as you mentioned).

**Comment by revelt** [comment]
> (not pre ES5 as you mentioned)

That's a fact: ES5 spec prohibits reassigning `undefined`. So, speaking strictly, `void 0` is necessity only for pre-ES5. Feel free to double-check the sources, I remember it was written in JS books too.

Your point about taking less space is valid. Maybe that's the way to go...

**Comment by Andarist** [comment]
> I don't think it actually hurts perf, void 0 is two operation but it's guarantee to always have the same result unlike undefined (not pre ES5 as you mentioned).

I'm wondering if that's even true (that it is an extra operation). Compiler should be smart enough to just recognize this is undefined and shouldnt penalize this, as this is such a common pattern.

**Comment by revelt** [comment]
That would be an interesting perf test actually.
